influxdb3_write/src/write_buffer/mod.rs

```diff
@@ -1647,6 +1669,91 @@ mod tests {
+            WalConfig {
+                gen1_duration: Gen1Duration::new_1m(),
+                max_write_buffer_size: 100,
+                flush_interval: Duration::from_millis(10),
+                snapshot_size: 1,
+            },
+        )
+        .await;
+
+        let db_name = "foo";
+        let tbl_name = "bar";
+
+        do_writes(
+            db_name,
+            &wbuf,
+            &[
+                TestWrite {
+                    lp: format!("{tbl_name},t1=a f1=true"),
+                    time_seconds: 1,
+                },
+                TestWrite {
+                    lp: format!("{tbl_name},t1=b f1=false"),
+                    time_seconds: 2,
+                },
+                TestWrite {
+                    lp: format!("{tbl_name},t1=c f1=true,f2=42"),
+                    time_seconds: 3,
+                },
+            ],
+        )
+        .await;
+
+        verify_snapshot_count(1, &wbuf.persister).await;
+
+        // There should be no parquet files being gotten from at this point, because we haven't
+        // even made any queries yet:
+        let counts = obj_store.get_counts();
+        println!("get counts: {counts:#?}");
+        assert!(counts
+            .iter()
+            .all(|(path, _)| !path.to_string().ends_with(".parquet")),);
+
+        // The same goes for get_range requets:
+        let counts = obj_store.get_range_counts();
+        println!("get_range counts: {counts:#?}");
+        assert!(counts
+            .iter()
+            .all(|(path, _)| !path.to_string().ends_with(".parquet")),);
+
+        let batches = get_table_batches(&wbuf, db_name, tbl_name, &ctx).await;
+        assert_batches_sorted_eq!(
+            [
+                "+-------+------+----+----------------------+",
+                "| f1    | f2   | t1 | time                 |",
+                "+-------+------+----+----------------------+",
+                "| false |      | b  | 1970-01-01T00:00:02Z |",
+                "| true  |      | a  | 1970-01-01T00:00:01Z |",
+                "| true  | 42.0 | c  | 1970-01-01T00:00:03Z |",
+                "+-------+------+----+----------------------+",
+            ],
+            &batches
+        );
+
+        let counts = obj_store.get_counts();
+        println!("get counts: {counts:#?}");
+        // There still should be no parquet files gotten yet, as the cache should have overridden
+        // any get request to the object store:
+        assert!(counts
+            .iter()
+            .all(|(path, _)| !path.to_string().ends_with(".parquet")),);
+        // However, now there are requests being made through get_range, presumably during schema
+        // inference on the parquet files by datafusion:
+        let counts = obj_store.get_range_counts();
+        println!("get_range counts: {counts:#?}");
+        assert!(counts
+            .iter()
+            .any(|(path, count)| path.to_string().ends_with(".parquet") && *count == 3));
```

Comment:
I stumbled upon this somewhat by accident, but it seems that Datafusion is making GET requests to infer schema of the parquet file at some point. It uses GET_RANGE requests to only get a portion of the actual file bytes, but I suspect this will still incur network overhead.

I am not sure I can bypass these requests with the way we are caching parquet files in this PR, since these requests are coming from Datafusion internals.

This would seem to be an argument in favour of the "decorator" style caching that IOx/core has.

Reply:
bah, yeah that'll defeat the point of caching if we're still making object store requests.

Reply:
The IOx team had worked on metadata caching in RAM, which this is likely what is getting looked up. I'm wondering if we can leverage what they've done there to make it so that we pass in that cached data for things that are in the cache?

Reply:
to avoid the initial fetch for parquet metadata, you need to override the default reader in DataFusion

The overall example is here: https://github.com/apache/datafusion/blob/main/datafusion-examples/examples/advanced_parquet_index.rs#L492-L495

The factory registration is here: 
https://github.com/apache/datafusion/blob/98e5f64df37b7a3a70ce5018d0d4098d0bed6ddb/datafusion-examples/examples/advanced_parquet_index.rs#L492-L495

And then the cached reader is here: https://github.com/apache/datafusion/blob/98e5f64df37b7a3a70ce5018d0d4098d0bed6ddb/datafusion-examples/examples/advanced_parquet_index.rs#L587-L591

there are a few configuration settings I found I had to set while working on the example that are noted

Reply:
I am not quite sure how this is hooked up internally -- I believe @erratic-pattern  connected it in InfluxDB 3.0 distributed -- https://github.com/influxdata/influxdb_iox/pull/11876 specifically -- so maybe he has some hints too

Reply:
Thanks @alamb - that is helpful I will have a read through!